src/platform/tests/TestDnssd.cpp

```diff
@@ -14,47 +35,73 @@ using chip::Dnssd::DnssdService;
 using chip::Dnssd::DnssdServiceProtocol;
 using chip::Dnssd::TextEntry;
 
-static unsigned int gBrowsedServicesCount  = 0;
-static unsigned int gResolvedServicesCount = 0;
-static bool gEndOfInput                    = false;
+namespace {
+
+struct DnssdContext
+{
+    nlTestSuite * mTestSuite;
+
+    std::atomic<bool> mTimeoutExpired{ false };
+
+    intptr_t mBrowseIdentifier = 0;
+
+    unsigned int mBrowsedServicesCount  = 0;
+    unsigned int mResolvedServicesCount = 0;
+    bool mEndOfInput                    = false;
+};
+
+} // namespace
+
+static void Timeout(chip::System::Layer * systemLayer, void * context)
+{
+    auto * ctx = static_cast<DnssdContext *>(context);
+    ChipLogError(DeviceLayer, "mDNS test timeout, is avahi daemon running?");
+    ctx->mTimeoutExpired = true;
+    chip::DeviceLayer::PlatformMgr().StopEventLoopTask();
+}
 
 static void HandleResolve(void * context, DnssdService * result, const chip::Span<chip::Inet::IPAddress> & addresses,
                           CHIP_ERROR error)
 {
+    auto * ctx   = static_cast<DnssdContext *>(context);
+    auto * suite = ctx->mTestSuite;
     char addrBuf[100];
-    nlTestSuite * suite = static_cast<nlTestSuite *>(context);
 
     NL_TEST_ASSERT(suite, result != nullptr);
     NL_TEST_ASSERT(suite, error == CHIP_NO_ERROR);
 
     if (!addresses.empty())
     {
         addresses.data()[0].ToString(addrBuf, sizeof(addrBuf));
-        printf("Service[%u] at [%s]:%u\n", gResolvedServicesCount, addrBuf, result->mPort);
+        printf("Service[%u] at [%s]:%u\n", ctx->mResolvedServicesCount, addrBuf, result->mPort);
     }
 
     NL_TEST_ASSERT(suite, result->mTextEntrySize == 1);
     NL_TEST_ASSERT(suite, strcmp(result->mTextEntries[0].mKey, "key") == 0);
     NL_TEST_ASSERT(suite, strcmp(reinterpret_cast<const char *>(result->mTextEntries[0].mData), "val") == 0);
 
-    if (gBrowsedServicesCount == ++gResolvedServicesCount)
+    if (ctx->mBrowsedServicesCount == ++ctx->mResolvedServicesCount)
     {
+        chip::DeviceLayer::SystemLayer().CancelTimer(Timeout, context);
+        // After last service is resolved, stop the event loop,
+        // so the test case can gracefully exit.
         chip::DeviceLayer::PlatformMgr().StopEventLoopTask();
```

Comment:
I've just found one issue with that....

When `DnssdResolveCallback` callback of `ChipDnssdResolve()` is called, the chip stack should be locked (there is a lock assert during message exchange - e.g. PASE session pairing) if the callback is called from non-matter thread. So, if `HandleResolve` is called from non-matter thread, the `StopEventLoopTask()` will actually wait from event loop to terminate (which will try to lock chip stack :D). This, popped out on Tizen.

Do not merge until this comment will be resolved!

Reply:
> So, if HandleResolve is called from non-matter thread

That is .. not an OK thing to happen.  `ChipDnssdResolve` assumes that it will have its callback called on the Matter thread, no?

Reply:
Ah, I guess that's true, you could be holding the stack lock instead of being on the Matter thread.  OK, I see.